pkg/kubelet/dockershim/docker_streaming_windows.go

```diff
@@ -0,0 +1,95 @@
+// +build windows
+
+/*
+Copyright 2019 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package dockershim
+
+import (
+	"bytes"
+	"fmt"
+	"io"
+
+	runtimeapi "k8s.io/kubernetes/pkg/kubelet/apis/cri/runtime/v1alpha2"
+	"k8s.io/kubernetes/pkg/kubelet/dockershim/libdocker"
+
+	dockertypes "github.com/docker/docker/api/types"
+	dockercontainer "github.com/docker/docker/api/types/container"
+)
+
+func portForward(client libdocker.Interface, podSandboxID string, port int32, stream io.ReadWriteCloser) error {
+	targetContainer, err := client.InspectContainer(podSandboxID)
+	if err != nil {
+		return err
+	}
+	if !targetContainer.State.Running {
+		return fmt.Errorf("target container (%s) is not running", podSandboxID)
+	}
+	metadata, err := parseSandboxName(targetContainer.Name)
+	if err != nil {
+		return fmt.Errorf("error parsing sandbox metadata: %v", err)
+	}
+	name := randomizeName(makeContainerName(&runtimeapi.PodSandboxConfig{
+		Metadata: metadata,
+	}, &runtimeapi.ContainerConfig{
+		Metadata: &runtimeapi.ContainerMetadata{
+			Name: "portforward",
+		},
+	}))
+
+	wincat, err := client.CreateContainer(dockertypes.ContainerCreateConfig{
+		Name: name,
+		Config: &dockercontainer.Config{
+			Image:        "gcr.io/cf-london-servces-k8s/windows-images/wincat:latest",
```

Comment:
Worth pointing out that this specifically shouldn't be merged, my idea is that this would be something like `"wincat"` and would be expected to be on the node already and not pulled from a registry, just as the pause image is today

Reply:
hm, but isn't the pause image still pulled on Windows nodes if it's not present?
it was recently added to the official MS registry for that purpose.

Reply:
I hadn't heard anything about that. In the [SDN code](https://github.com/Microsoft/SDN/blob/81b1ab7cda419ed54a4b36c497b4414bafbf5f91/Kubernetes/windows/start-kubelet.ps1#L118) the pause container is still being specified as `kubeletwin/pause` which doesn't appear to be on Dockerhub

Reply:
Looks like it's [the same](https://github.com/Azure/aks-engine/blob/09564f98115912a60144771e27b0c1fb2ef3aa77/parts/k8s/windowskubeletfunc.ps1#L133-L157) for aks-engine

Reply:
this is the image that i saw (+tested):
`mcr.microsoft.com/k8s/core/pause:1.0.0`
(ML 1803/1809, nanoserver based)

https://github.com/kubernetes/kubernetes/issues/74218#issuecomment-472085294

Reply:
if that's the direction we go in for pause then we could do something similar for this image. i'd need to introduce code to pull the image if it's not present.

Reply:
+1 of the pause container. @PatrickLang Have we included the wincat.exe in the pause image?

Reply:
Good point, we could just use the pause container and `exec` wincat inside it if it's already inside

Reply:
Packing wincat.exe in the pause image seems like a reasonable workaround. Question: how does this affect the resource usage and the stability of the pause container?

I don't think we need to build everything on the node, even though it may be faster. User/vendors can preload the images if they want. Otherwise, distributing the source to nodes become another issue.

BTW, whatever it's decided here may be used by the containerd integration as well.

Reply:
Also, the equivalent of pod-level cgroup has not been implemented on Windows, right, @feiskyer @PatrickLang? Just wanted to know how this may affect the resource isolation/accounting in the future.

Reply:
> Also, the equivalent of pod-level cgroup has not been implemented on Windows, right, 

Right.